### ISSUE TITLE
Improve documentation / faq to cover these points

### ISSUE THREAD

**Issue by andrewcharnley** [lerna/lerna#1788]
Hi,

I've upgraded from Lerna 2 to 3. 

**Issue 1** - while bootstrap --hoist is good in principle I found many newer npm commands (outdated, audit) won't work. In addition, with --hoist no package-lock.json is created in each folder - which is surely detrimental when it comes to publishing packages that have dependencies (non dev)? One option is to remove semver but I'd prefer to wipe all package-lock.json files to update to minor/patch dependency revisions.

**Issue 2** - the new lerna link command is not well explained, especially _lerna link convert_. Rather than hoisting I've used this command and I can see file:// references within each package.json. Dilemma is when published these locations won't point to them. Does lerna swap them out before publish?

The good news is without --hoist the package-lock.json's return.

**Issue 3**
Since all the devDependencies have moved to the root (which I agree with since I have several packages all using the same devDeps) I can no longer work directly in an individual folder. For example if I dive into package/x and do npm test and within the script->test is "mocha" then mocha will no longer be found. I guess it's because npm tries the node_modules/.bin folder but doesn't traverse through parent folders (possibly a bug of npm imo). Indeed if I switch the mocha command to be ../../node_modules/.bin/mocha it will work. Thing is, I have a lot of commands (babel, documentation, flow) in a lot of packages and it doesn't feel clean. Finally I tried npx mocha - worse as it attempts to install mocha into the current directory.

I'm not saying it's a bug of lerna, indeed _lerna run_ is working correctly but I do feel _lerna link convert_ should be documented. There are times when I don't want to use lerna run and --scope to run a command, such as when running an individual test within a package.

Thanks,

Andrew

**Comment by evocateur** [comment]
> when published these locations won't point to them. Does lerna swap them out before publish?

Yep. All local `file:` specifiers are temporarily resolved to the version range of the package they point to before being packed into the tarball. Lerna itself uses this pattern.

Yes, the pattern introduced by `lerna link convert` should be better-documented. For the time being, all I can point to is the source of lerna itself and a couple other public repos (https://github.com/zillow/javascript, https://github.com/evocateur/pectin, etc)

> Issue 3

Generally speaking, a monorepo's tests should be runnable from the root. It is up to the test runner to be flexible enough to run focused tests as desired (this is one reason among many that `jest` far outstrips `mocha`, imo). I suppose `lerna exec --scope=my-package-name -- mocha` would work for you, since `lerna exec` adds the root `node_modules/.bin` to the PATH when executing.

**Comment by revelt** [comment]
> In addition, with --hoist no package-lock.json is created in each folder - which is surely detrimental when it comes to publishing packages that have dependencies (non dev)?

@andrewcharnley it depends — if lerna packages are npm libraries, then it's fine not to have `package-lock`s. But if your project is web app parts then yeah, possibly. There was a good [discussion](https://github.com/sindresorhus/ama/issues/479#issuecomment-310661514) leaning against `package-lock` in npm libraries.

Personally, I tackle **issue 1** this way. I wrote a node CLI app which reads all `package.json` within `packages/*` and pings npm and writes the latest fetched version, with `^` prepended. This might be impossible on web app projects but does work on npm libraries. The API surface is small-enough, it's extremely unlikely that non-major dependency releases would break unit tests. Even major releases can be bumped 90% of the times from my experience. Again, this is definitely not the case if anything `express` or `react` related is among dependencies (imho).

Personally, I never experienced **issue 2** because I use `lerna bootstrap --ignore-scripts --hoist`. But yeah, it's not clear in the documentation when should it be used and when not.

With regards of issue 3. It might be or not be applicable in your case but I'm running tests on `nyc` + `ava` and I don't have them installed globally and the following script on each package works fine. I `cd` into one of `packages/*` folders (npm libraries), do all the development on them and run tests fine, as usual:

```
"test": "./node_modules/.bin/eslint src/*.js test/*.js && nyc ava"
```

Notice how above I'm calling `eslint` with a relative path and `nyc` directly. Both work fine. All my bin CLI scripts got sym-linked: `ava`, `eslint`, `nyc`, `prettier`, `rollup`.. Maybe try wiping everything (`rm -rf node_modules && npx lerna clean -y`) and `lerna bootstrap --ignore-scripts --hoist`? Btw, my own repo is [this](https://bitbucket.org/codsen/codsen/src/master/), maybe you'll find something handy in `scripts`. I'm combatting conventional changelogs at the moment, but hoisting, running scripts on packages and dependencies are "nailed" 100% on lerna setup side.

**Comment by estaub** [comment]
> the new `lerna link` command is not well explained, especially `lerna link convert`

This is very nearly a show-stopper for me; I'm trying lerna for the third time, in large part to get hoisting, and AFAICT there is not one word anywhere about `lerna link convert`, even though it's described as the standard way to hoist.  If it isn't obvious, I/we need to understand: 

- what gets hoisted and what doesn't
- behavior on version conflicts - does one get hoisted & if so, which?
- how to prevent hoisting of specific packages
- techniques for ensuring common versioning (and hence hoisting)  of certain dependencies and devdependencies across packages.  I suspect use of a base package with peerDependencies will enforce for normal dependencies, but I don't see a way to do it for devDependencies
- behavior when publishing: what "unhoisting" occurs?

That's just off the top of my head.  Some of these may be moot with even a rudimentary understanding of what `convert` does, but that's my point.  You may consider the "techniques" item to be a bit out of scope, but I think it would be of great value to lerna learners even if not about the specific behavior of lerna itself.

**Comment by andrewcharnley** [comment]
Hi,

lerna bootstrap --hoist is defunct when using link convert as the child packages have all their devDependencies moved into the parent package.json while the child packages are listed as dependencies on the parent package.json so the child dependencies are installed. It's rather nice!

estaub -> use lerna link convert if you're using a recent npm version (whichever supports symlinks, I think it's v5+). After this your child packages won't have any devDependencies in them and keep it that way in the future by adding them to the root package.json. If you add new packages be sure to add them to the dependencies list in the root package.json (you'll notice lerna link convert has done it for you for the ones you already have). This makes sense because a published package doesn't need/use devDependencies.

After that you only need to run npm i at the root level, no bootstrapping or hoisting. When it comes to publishing, Lerna swaps out the link strings to reference the package name & version. 

Cheers, Andrew

**Comment by andrewcharnley** [comment]
PS) I couldn't get nyc, mocha etc to run within a child folder without forcing a relative path to the parent dir with ../../.bin. npx can't find it either so it doesn't like traversing up the tree to find binaries. I can live with it just fine though.

**Comment by evocateur** [comment]
https://github.com/lerna/lerna/issues/1419#issuecomment-456182409 indicates we probably have an existing problem with lerna when it runs package scripts attempting to use binaries installed up the tree...

**Comment by TheSpyder** [comment]
> Yep. All local `file:` specifiers are temporarily resolved to the version range of the package they point to before being packed into the tarball. Lerna itself uses this pattern.

I really wish this was documented better (I haven't found any reference to it yet other than this ticket). I am experimenting with migrating a large project to a monorepo and spent a bunch of time setting up a script to `sed` the incoming package.json files because it looked like `lerna link convert` wasn't giving me what I needed 😳

I'm also seeing yarn warnings about colliding ranges as a result of running `convert`, I'm still determining whether I've done something wrong or it's an issue with the command.

**Comment by evocateur** [comment]
If you're using Yarn you might as well use Workspaces. Notably, Yarn does not support relative `file:` specifiers, they chose `link:` (Lerna works with both, although again, you might as well use yarn workspaces).

**Comment by TheSpyder** [comment]
I am using workspaces, but I get the impression lerna is still valuable for things like publishing. Mostly I was looking for the behaviour to hoist the `devDependencies` references to the root, yarn auto-hoisted most things but not all so I figure specifying them in the root would be a better place to resolve minor differences.

Thanks for pointing me to `link:`, that helps a lot :)

I'm fairly sure what I see is a bug, I'll work on a reproduction case and file a new issue. `lerna link convert` adds my in-repo packages to the root `package.json` `devDependencies` as `file:../pkgname` in addition to the `dependencies` `file:packages/pkgname`.

**Comment by andrewcharnley** [comment]
> 
> I'm fairly sure what I see is a bug, I'll work on a reproduction case and file a new issue. `lerna link convert` adds my in-repo packages to the root `package.json` `devDependencies` as `file:../pkgname` in addition to the `dependencies` `file:packages/pkgname`.

This is the correct logic. I'm using npm v6 - don't see a need for yarn anymore - so can't comment on that side of it. At this point all Lerna is doing is iterating the folders and running commands in them (plus a bit of version management and publishing spice if you want it).

**Comment by TheSpyder** [comment]
The version management and publishing is what I'm looking for, I'm still experimenting with how I want the monorepo to work for us. Even if it's the expected behaviour based on what the convert script is doing, finishing with file links that point outside of the monorepo is a fairly clear bug.

**Comment by andrewcharnley** [comment]
I don't have that issue at at. The file links in parent/children package.json(s) resolve correctly from parent to children and child to parent.

**Comment by TheSpyder** [comment]
I realise that, hence my plan to create a reproduction case and report as a bug. Sorry for dragging the thread off topic :)

**Comment by krumware** [comment]
Piggybacking on the devDependency hoisting discussion: I wasn't able to find a good answer related to `lerna link convert`'s hoisting and removal of the common devDependencies packages' package.json. I'm trying to focus on what someone may suggest as "best practice" (big air quotes).

After running `npm link convert` which removes shared dev dependencies from relevant packages' package.json:
We know that running an npm command in a package, perhaps a test run, ideally would resolve up and run successfully. This already works great when leveraging lerna to essentially run a test suite. But what if a developer or automated system now wants to run a `npm run test` in a single package, without installing all of the root's dev dependencies? Or if one of the packages is included in another project without a --production flag, and those dev dependencies are needed for whatever purpose? In either of those cases, without the package's devDependencies included, those dependencies won't be resolved.

It is likely wayyy too much effort to track and re-add those dependencies, like the convenient publish-time version inclusion for local dependencies, I don't want to propose anything like that.

But with the all of the above in mind, is it ok for us to re-add and keep those devDependencies in place in each package after the initial conversion, or will that cause problems for lerna? (assuming the dev or system is smart enough not to run `npm install` everywhere and cause problems on their own). I have tried this locally, and everything seems to be in order, but it wouldn't suprise me if there is behavior I don't know to expect.
Is there an approach that would be good to keep in mind for project composure?

**Comment by mchelen** [comment]
Can someone explain how the non-dev dependencies in leaf packages are handled after `lerna link convert`? Wouldn't you still need `lerna boostrap` to get those installed?

**Comment by andrewcharnley** [comment]
Not with npm v6, with npm i, when the dependency in package.json starts with "file:" it dives into it

**Comment by evocateur** [comment]
> I have a lot of this going on in my root package.json...

If you have lerna in your root `devDependencies`, you certainly don't need to prefix every script execution of `lerna` with `npx`. Anyway, nobody should be using `lerna bootstrap` when Yarn workspaces, `pnpm recursive`, and acceptable(ish) workarounds for `npm` exist.

**Comment by pleger-cdpq** [comment]
@andrewcharnley for the problem you are describing here:

> PS) I couldn't get nyc, mocha etc to run within a child folder without forcing a relative path to the parent dir with ../../.bin. npx can't find it either so it doesn't like traversing up the tree to find binaries. I can live with it just fine though.

you can run `link-parent-bin` as a `postinstall` step which will link the dev .bins into the child packages

https://www.npmjs.com/package/link-parent-bin

@mchelen like @andrewcharnley said you just have to put your packages as dependencies, e.g.:

```json
"dependencies": {
    "@ai-ui/core-docs": "file:packages/core-docs",
    "@ai-ui/core-ui-lib": "file:packages/core-ui-lib",
    "@ai-ui/core-webpack-config": "file:packages/core-webpack-config",
    "@ai-ui/core-yo-generator": "file:packages/core-yo-generator"
  },
```

**Comment by pleger-cdpq** [comment]
@andrewcharnley 

>As such Lerna doesn't have as much benefit as it used to. Personally I find it's just a means of running commands in parallel across all packages. I say this having not fully utilised the versioning management which I'm sure is the other benefit.

I got to the same conclusion.

where I still like lerna is to handle automated changelogs, i.e. with:

```bash
lerna version --conventional-commits --create-release github
```

**Comment by char0n** [comment]
> Not with npm v6. Using "npm i" when the dependency in package.json starts with "file:" it dives into it. Magic.

@andrewcharnley the "diving" only works for `dependencies` but not `devDependencies` in leaf packages. This mean all `devDependencies` of all packages gets mixed into root `devDependencies`. Even those that are really specific to only particular leaf package. Is there a way how to solve this? 

Thanks!

**Comment by wolfhoundjesse** [comment]
> Anyway, nobody should be using `lerna bootstrap` when Yarn workspaces, `pnpm recursive`, and acceptable(ish) workarounds for `npm` exist.

On the front page of https://lerna.js.org/, it says `lerna bootstrap` is crucial. Are you saying it's no longer the case as of npm@6?

**Comment by char0n** [comment]
I've compiled an [article](https://www.linkedin.com/pulse/things-i-wish-had-known-when-started-javascript-monorepo-gorej/) using information from this issue, various different issues and my personal experience with lerna after introduction of npm local paths. Hope it helps with confusion about how to use lerna nowadays.
